### PR TITLE
Fix language selector not scrolling with keyboard

### LINKED ISSUE #28345: Language picker doesn't scroll down when using arrow keys
## What
When you navigate to your "Account settings" and change the "Language" settings the picker component doesn't scroll and show more options when you navigate down through options with arrow keys. [Loom with more context](https://www.loom.com/share/b87d38299a8648479fe378282563ae1d).

## Expected Behavior
It should behave like other picker components, and scroll down when using arrow keys to browse languages.

### PR DESCRIPTION
**Pull request by EmmadUsmani**
Closes https://github.com/metabase/metabase/issues/28345

### Description

The language picker in the account settings page was not scrolling when the user selected a non-visible option with their keyboard. The problem was that the `Popover` component was not passing the `maxHeight` prop to `AccordionList`. I fixed this by refactoring the code in `Popover` to make sure `childProps` were being given in the case that `this.props.children` was an array of multiple children, which was not being done previously.

_Note: this fixes the problem for many other use cases of the `AccordianList` component. However, some uses like the filter picker in the notebook editor still do not scroll properly. I will file a separate issue to track that._

### How to verify

Describe the steps to verify that the changes are working as expected.

1. Settings -> Account settings -> Language
2. Try navigating through select menu with keyboard, confirm it scrolls

### Demo

https://user-images.githubusercontent.com/37751258/221035626-464fa64d-9aca-4910-921e-17afc43721bd.mov

### Checklist

- [✅] Tests have been added/updated to cover changes in this PR

### CODE REVIEW

**Comment by gusaiani** [comment]
Hi Emmad, thanks for the PR.

I will manual test this as well as I can a bit later today, and encourage others to do the same.

Popover has a multitude of use cases and regressions are common.

Just to start, we should look at filtering in questions, usage inside sidebars.

**Comment by EmmadUsmani** [comment]
> Hi Emmad, thanks for the PR.
> 
> I will manual test this as well as I can a bit later today, and encourage others to do the same.
> 
> Popover has a multitude of use cases and regressions are common.
> 
> Just to start, we should look at filtering in questions, usage inside sidebars.

Thank you Gus, that sounds good to me! It would be great if you could list out places to manually test in a comment, so myself and others can reference them.

**Comment by gusaiani** [comment]
@EmmadUsmani 

>  It would be great if you could list out places to manually test in a comment,

Yup, I've started it here:
https://www.notion.so/metabase/Popover-Manual-Exploration-Checklist-7c6d653de998445d8da16c2f8b38ba15

Haven't even scratched the surface, please add to it if you can.

**Comment by WiNloSt** [comment]
> Thank you Gus, that sounds good to me! It would be great if you could list out places to manually test in a comment, so myself and others can reference them.

😬 that's gonna be a lot. @EmmadUsmani. Usually, I use TypeScript (aka <kbd>Shift</kbd> + <kbd>F12</kbd> on VS Code) to check where Popover would be used.

See the scrollbar on the right of the screenshot? I think you'd get an idea where this is going.
Screenshot 2023-02-24 at 12 48 07 PM

**Comment by nemanjaglumac** [comment]
> I will manual test this as well as I can a bit later today, and encourage others to do the same.

Thank you for doing this, @gusaiani 
May I suggest we don't switch context and keep everything here in a comment. It might be useful in the future for other devs, and Notion page might be deleted.

**Comment by gusaiani** [comment]
@EmmadUsmani 

Happy to paste the checklist here.
Thing is, we touch Popover a lot.
So in the future this checklist, long overdue imho, will be needed.
Redundancy may be good here :)

File: frontend/src/metabase/components/Popover/Popover.jsx [hunk 1]
```diff
@@ -240,14 +240,11 @@ export default class Popover extends Component {
...
-            // NOTE: workaround for https://github.com/facebook/react/issues/12136
-            !Array.isArray(this.props.children)
-          ? React.cloneElement(
-              React.Children.only(this.props.children),
-              childProps,
-            )
-          : this.props.children}
+          : React.Children.map(this.props.children, child =>
+              React.isValidElement(child)
+                ? React.cloneElement(child, childProps)
```
Comment (WiNloSt) [inline frontend/src/metabase/components/Popover/Popover.jsx:245]: This issue seems to be caused by https://github.com/metabase/metabase/pull/25708. Historically, we didn't think it gonna cause any issues. However, due to the fact that we need a custom footer on the map options


currently, we pass along `footer` together with the `Accordion` that's why it doesn't get the `maxHeight` prop.

https://github.com/metabase/metabase/blob/ad8a8982ec3f053dd156b13c16560c539739a33e/frontend/src/metabase/core/components/Select/Select.tsx#L325

However, in that case, we'll also pass along `maxHeight` to the footer which doesn't make sense.
image

I manually check every place used by `Popover` and only 3 have passed more than 1 child.

1. DataSelector: Although, keyboard arrows don't work, <kbd>Tab</kbd> navigation works fine.
    image
2. TimeseriesFilterWidget: It's so short it shouldn't be a problem.
Screenshot 2023-02-24 at 1 32 31 PM
3. Select:  which this PR attempts to fix

I proposed you worked with @npfitz to come up with an API that will fix the `Select` component that allows it to pass additional `footer` elements that won't break the accordion `maxHeight` prop. I've try to see if it's possible to use `sections` props in `AccordionList` to receive the `footer`, but seems `sections` expect `items` key which would be selected by users. However, I didn't give it a more thorough investigation.

**Review by WiNloSt — CHANGES_REQUESTED** [review]
I think most of this fix will account for only a single place, and as I have mentioned, I don't think it makes sense to pass `maxHeight` to every child in there, as all children combined should equal `maxHeight`. i.e. I think it still makes sense to only pass `maxHeight` when there's a single child inside the popover.

I think we should come up with a better API to accommodate https://github.com/metabase/metabase/pull/25708 while not causing this bug.

What do you think

**Comment by gusaiani** [comment]
@WiNloSt, just so I stay in the loop, or maybe drop out of this review:

I tried to retrace the steps for cases you mention in the screenshots above, and they look ok afaict.

So far, have you detected any regression stemming from the changes made here?

**Comment by WiNloSt** [comment]
> @WiNloSt, just so I stay in the loop, or maybe drop out of this review:
> 
> I tried to retrace the steps for cases you mention in the screenshots above, and they look ok afaict.
> 
> So far, have you detected any regression stemming from the changes made here?

@gusaiani No, this PR fixes the issue and doesn't introduce any regression as far as I can tell. However, I don't think the logic is correct.
1. It suggests that we should pass `maxHeight` to every child
2. As I have mentioned there's only a handful of components that has more than 1 child. And this whole change seemingly only fixes a single component behavior `Select`. Given that I think it might be better to fix `Select` directly instead.

**Comment by gusaiani** [comment]
> No, this PR fixes the issue and doesn't introduce any regression as far as I can tell. However, I don't think the logic is correct.

Gotcha @WiNloSt, thanks.

The interesting thing, orthogonally to Select, is observing that the code Emmad changed seems simpler than what we currently have.

This leads me to lightly think that we could use this fix as it may also handle future variations.
Maybe in addition to any tweaks to Select.
I could be wrong here, just expressing my thinking for now.

**Comment by WiNloSt** [comment]
> > No, this PR fixes the issue and doesn't introduce any regression as far as I can tell. However, I don't think the logic is correct.
> 
> Gotcha @WiNloSt, thanks.
> 
> The interesting thing, orthogonally to Select, is observing that the code Emmad changed seems simpler than what we currently have.
> 
> This leads me to lightly think that we could use this fix as it may also handle future variations. Maybe in addition to any tweaks to Select. I could be wrong here, just expressing my thinking for now.

@gusaiani I agree @EmmadUsmani 's code looks simple, but I don't think it's correct. Technically, it works perfectly fine by fixing the regression created by https://github.com/metabase/metabase/pull/25708, but simple doesn't always correct.

As I've explained. The API Popover has right now is saying if we have 1 child we're gonna pass `maxHeight` to the child then it could do whatever it wants with this information, so it works nicely with the popover. However, that would be problematic when we say we will pass `maxHeight` to every single child. This API just simply looks wrong. I couldn't imagine how each child will coordinate on the `maxHeight` they received, says we have 3 children, how are they going to work together to make sure that they won't set their heights in combination exceeding the `maxHeight`. The only correct API I see here is to pass `maxHeight` to a single child (the current behavior).

The tl;dr is that. 1. Yes this code looks simple, but 2. the API looks wrong, and we shouldn't trade correctness for simplicity. Hope that makes sense.

**Review by alxnddr — COMMENTED** [review]
I looked through all usages in the app and I tend to agree with @WiNloSt. 

To have a bulletproof fix we need to deal with the Select footer that we use here:

**Comment by gusaiani** [comment]
@WiNloSt @alxnddr @EmmadUsmani 

Awesome, thanks.
I'll keep manual exploration of Popovers at pause for now.

**Comment by EmmadUsmani** [comment]
Thanks for the thoughtful reviews everyone. I agree with @WiNloSt, it doesn't make sense for all children to receive the `maxHeight` prop, it definitely would make more sense to fix the `Select` component directly.

In the interest of time I am going to drop this for now, but may revisit in the future.
